### ISSUE TITLE
Strange error

### ISSUE THREAD

**Issue by vdbergh** [glinscott/fishtest#1156]
This error is is caused by a problem with the net

https://tests.stockfishchess.org/actions?count=1&before=1636978729.909
```
info string ERROR: If the UCI option "Use NNUE" is set to true, network evaluation parameters compatible with the engine must be available.
info string ERROR: The option is set to true, but the network file nn-b97fa10a6be6.nnue was not loaded successfully.
info string ERROR: The UCI option EvalFile might need to specify the full path, including the directory name, to the network file.
info string ERROR: The default net can be downloaded from: https://tests.stockfishchess.org/api/nn/nn-b97fa10a6be6.nnue
info string ERROR: The engine will be terminated now.
```
However then SF quits with exit code 1 (tested in a shell). Yet this was not caught by the worker (which tests for nonzero return codes).
```
        if p.returncode:
            raise WorkerException(
	        "Bench exited with non-zero code {}".format(p.returncode)
            )

```

**Comment by ppigazzini** [comment]
Commuting now, IMO that code need indentation to be under the context manager.

**Comment by ppigazzini** [comment]
On the smartphone I read the code in the wrong place.
The subprocess returncode works also out of Popen context manager.
```python
import subprocess
with subprocess.Popen(["ls", "non_existent_file"]) as p:
    p.communicate()
    print(f"inner: {p.returncode}")
print(f"outer: {p.returncode}")
```
```
python3 test_returncode.py
ls: cannot access 'non_existent_file': No such file or directory
inner: 2
outer: 2
```

**Comment by vdbergh** [comment]
> Commuting now, IMO that code needs indentation to be under the proper context manager.

Actually I think it has the right level of nesting because it is invoked with `with ExitStack() as stack:`. Unless I am misunderstanding this construction.

**Comment by vdbergh** [comment]
Actually now I think you are right. The return code should be checked _outside_ the `with ExitStack() as stack:` block. Thanks!

**Comment by ppigazzini** [comment]
Yes, reading on the smartphone I confused it with this code:
https://github.com/glinscott/fishtest/blob/fce3cc794ce8e9045559d512f40f2f7bc0c0b1c8/worker/games.py#L138-L150

But the info seems to be there also after the context manager-

**Comment by vdbergh** [comment]
> But the info seems to be there also after the context manager.

I think it is after you invoke `p.communicate()` (which reads everything, closes the fds and waits for the process to quit). But we don't use `p.communicate()` in the worker.

**Comment by vdbergh** [comment]
A bit orthogonal to the problem at hand but I don't see a compelling reason here to use `ExitStack`.

**Comment by vdbergh** [comment]
I will make a PR later. I think I will introduce a `RunException` which means stopping the run.

**Comment by vdbergh** [comment]
> A bit orthogonal to the problem at hand but I don't see a compelling reason here to use `ExitStack`.

Ah I see. The logic is a bit complex because of the `busy_process`.

**Comment by ppigazzini** [comment]
> A bit orthogonal to the problem at hand but I don't see a compelling reason here to use `ExitStack`.

I used it to deal with the condition::(https://github.com/glinscott/fishtest/blob/fce3cc794ce8e9045559d512f40f2f7bc0c0b1c8/worker/games.py#L258-L269)

**Comment by vdbergh** [comment]
Yes I see. So an `ExitStack()` object can have multiple enter contexts?

**Comment by vdbergh** [comment]
> Yes I see. So an `ExitStack()` object can have multiple enter contexts?

Yes it seems so according to the docs.

**Comment by ppigazzini** [comment]
Yes, `ExitStack()` is used when you have multiple and dynamic enter contexts.

**Comment by vdbergh** [comment]
Hopefully fixed by PR #1157 .
